rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,519 @@
+
+### References to the problems above
+
+This curated link collection highlights the Nix ecosystem's inconsistencies, a primary focus of this RFC.
+
+#### Nixpkgs - comment examples
+
+- [lib/attrsets](https://github.com/NixOS/Nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/Nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/build-support/trivial-builders/default.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/Nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/Nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/nixos/lib/make-disk-image.nix)
+
+# Design
+[design]: #detailed-design
+
+In the following, we give a comprehensive overview of the decisions that we've made. Detailed arguments for and against every decision can be found in the [Decisions](#Decisions) section
+
+## CommonMark
+
+**CommonMark according to [RFC 72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md) is the content of all doc-comments.**
+
+Adopting CommonMark as the content for all doc-comments brings the benefits of widely accepted and understood documentation format in tech projects while maintaining profitability and consistency within the Nix ecosystem by aligning with existing [NixOS/RFC-72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+
+## `/** */` is the doc-comment format
+
+The decision to use /** to start a doc-comment ensures a unique distinction from regular comments while still allowing seamless writing without IDE or editor support. A single choice provides the best developer experience and simplifies tooling.
+
+## Placement
+
+**The placement describes the relationship between doc-comments and the documentable node.**
+
+A **documentable node** can be:
+
+- Expression 
+- Binding 
+- Lambda Formal
+
+The following rules apply in descending order of precedence:
+
+- Doc-comments are placed before the **documentable node**. Only whitespace or non-doc comments are allowed in between. ([Examples](#basic-examples)) 
+
+- The documentation present before the `attribute path` describes the body of the attribute. ([Examples](#Attributes))
+    - In case placement is ambiguous, the one closer to the body has higher precedence. ([Examples](#ambiguous-placement))
+
+### Examples
+
+#### Basic examples
+
+Only whitespaces between the `documentable node` and the `doc-comment`
+
+```nix
+/**Doc for anonymous lambda function*/
+↓
+x: x;
+```
+
+```nix
+listToAttrs [
+  { name = "foo"; value = /**Documentation for '1'*/1; }
+]
+```
+
+#### Attributes 
+
+It is allowed to write the documentation before the attribute instead of placing it right before the body.
+
+```nix
+/**Doc for lambda function bound to a variable*/
+           ↓
+assigned = x: x;
+```
+
+```nix
+{
+    /**This documents the specialisation `map (x: x)` */
```

Comment:
Agreed, the  "only" thing we need is proper position tracking for some kind of values and expressions, in a very simplified model. Possible implementation could be to have it behind a cli-flag if it has major performance impacts. Generating documentation would be less performant, but at least beeing automated. We still have to explore the space of possible implementations to find the best one.

> If we'd want to support something like that, we'd need to make every attribute set expression essentially create a copy of every value.

Note that we didn't specify that values are documentable. The relation between expression and value is not very clear in some cases.

```nix
{
  /**This will show up */
                  ↓
  myFun = import ./my-fun.nix;

# The expression is:  ' import ./my-fun.nix; '
```

Reply:
> One problem with this approach is that it won't work for pure renames/aliases, e.g.

Also other optimizations, like beta-reduction-under-lambda and constant propagation.

> have it behind a cli-flag if it has major performance impacts.

It certainly will; it impedes a lot of optimizations.  Many of these are optimizations that cppnix doesn't/can't/won't implement, but other implementations will.  So the measurement of "performance impacts" can't be cppnix-specific.

I also think the cli-flag should inhibit the creation of derivations.  I cannot imagine any legitimate need for derivations to be influenced by docstrings.  This ensures that we can always disable docstring-reflection in the performance-sensitive builds without changing the outcome.

Reply:
> Note that we didn't specify that values are documentable.

That is a problem for dynamic documentation. There is only ever one expression, the top level expression (which won't be documented since it is user-supplied), when evaluating. Everything else will only be given as a value.

Reply:
> There is only ever one expression, the top level expression (which won't be documented since it is user-supplied), when evaluating.

I think we might have some misunderstanding on expressions.

"Expressions" are "eval"-able language constructs. For example, 


These things are expressions:

`1`
`1 1 1 1`  (This is a lambda "apply", (Call 1 (Arg 1 1 1))


But these things are not:

`...` -- ellipsis, part of lambda formal
`a = ` -- symbol, and the `=` in attrset

> There is only ever one expression

Hmm, I think we do have some nested expressions while eval?

e.g.

```nix
{ }: sub-expr
# ~~~~~~~~~ lambda
```

> That is a problem for dynamic documentation

Sorry, could you please elaborate what is "dynamic documentation"?

Reply:
I know what an expression is. You are missing some context: During discussion, implementability was a big point of discussion. Specifically the following ways of using the doc comments were discussed, also w.r.t. parity between them:

* Static doc generation, i.e. what nixdoc does right now
* Dynamically accessing doc comments, e.g. from REPL via the `:doc` command or from an evaluation via `builtins.unsafeGetDocComment` or similar.

The problem being, of course, that while during static doc extraction we are actually dealing with expressions, during evaluation there only ever is values. How these relate to the doc comments next to expressions is an important question to solve.

> Hmm, I think we do have some nested expressions while eval?

Of course, but `:doc lib.map` has one top level expression it recursively reduces to a Value—what doc comment next to which expression should be used? In that case you can come up with a cheap solution, of course, but values can arbitrarily be passed around, rebound and manipulated. Already `lib` is non-trivial, since `lib.map` and `lib.lists.map` are “the same”, albeit not equal, and can have differing doc comments next to the expression that “constitutes” them in each case.

I'm a bit confused that dynamic documentation is construed as a non-issue of RFC 145 at this point, as I remember it was one of the primary motivations of @hsjobeki's to undertake it in the first place. He was (is?) of the opinion that only dynamically introspecting `lib` using an algorithm written in Nix and extracting the attached documentation could lead to correct generated docs. In fact, he is already [using this approach for `docnix`](https://github.com/nix-community/docnix/blob/7aa0d59959f300815e8c10d71078fac65e548587/code-docs/mkDocs.nix) via a customized Nix implementation.

It is of course valid to only prescribe syntax for doc comments, but this should be made very explicit and incurs the consequence that it is more or less arbitrary what doc comment documents what “thing”. The potential future problem we are buying is that we can't replicate the same documentation output on the REPL dynamically, that we are generating statically for the nixpkgs manual, for example.

Reply:
Valid point. I am using dynamic documentation generation. This means I evaluate the nix expression tree of e.g. `.lib`
Then find all the lambdas and all doc-comments of them.
One of the biggest challenges is finding documentation for partially applied functions and their aliases. 
When I utilize the current lambda position tracking (with my custom nix version) and the already existent `unsafeGetAttrPos` I think we can build highly accurate documentation. 

With the RFC in its current state, I am very optimistic, the scope is very narrow, to just having the format and placement changed/adapted.

I think from the current state i or we could build a tool that serves as a genreic documentation renderer / extractor for nix projects.

A native implementation in the evaluator is currently moved out of scope. 

Of the RFCs scope: We should maybe move `unsafeGetAttrPos` behind a CLI flag and also add `unsafeGetLambdaPos` such that i don't have to maintain a separate fork of nix.

Reply:
> This means I evaluate the nix expression tree of e.g. `.lib`

Which is of course only ever given in value form.

> We should maybe move unsafeGetAttrPos behind a CLI flag

Probably hard to do, as it is required for nixpkgs' `meta` sets. We could probably drop it if `pkgs/by-name` is adopted more, but it is kind of essential so that `nix edit` can work.